### PR TITLE
Feature/double buff seq

### PR DESCRIPTION
**Pull request by MJGaughran**
This PR uses double buffering to improve the pandaseqtriggerpart.

As we only process the points in batches, the initial configuration time is reduced for long scans. In addition, as there is no limit on the total number of sequencer rows used during a scan, there is no upper limit on the total scan duration.

A variation of this PR was tested by Tom Trafford and appeared to work correctly. I still need to finish some changes to get scanning to work in the case where no detectors are selected.

Unit testing has also been reworked, as the previous method did not allow for proper testing of subscription-based logic.

Please note: I have made a minor change to the double_seq_pcomp block design, so if anyone is using a custom block design they will need to modify it accordingly. You can either look at the diff between versions, or load the updated double_seq_pcomp on top of your existing design.

### CODE REVIEW

**Comment by MJGaughran** [comment]
@BenBradnick @tomtrafford Do you find that you get issues with scans with no detectors, regardless of the changes in this PR?

If so, this pull request is not a WIP, and we can merge it after testing on a beamline.

**Review by BenBradnick — COMMENTED** [review]
Looks good to me, except that perhaps a list is not the clearest structure for storing all of the row data.

File: malcolm/modules/ADPandABlocks/doublebuffer.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+
+            if self._rows[-1][0] > 1:
+                final_row = list(self._rows[-1])
+                final_row[0] -= 1
+                self._rows[-1] = tuple(final_row)
+                final_row[0] = 1
+                self._rows += [tuple(final_row)]
+
+        final_self_row = list(self._rows[-1])
+        final_self_row[10] -= SEQ_TABLE_SWITCH_DELAY
```
[the remark this answers is not in the document]
Comment (MJGaughran) [reply malcolm/modules/ADPandABlocks/doublebuffer.py:100]: I'm going to resolve this, as it effectively duplicates your other comment about row indices being unclear.

File: malcolm/modules/ADPandABlocks/parts/pandaseqtriggerpart.py [hunk 2]
```diff
@@ -26,57 +23,250 @@
...
-        0,
-        0,
-        0,
-    ]
-    return row
+# SeqTriggers processing batch size
+BATCH_SIZE: int = 5000
+
+
+class SeqTriggers(object):
```
[the remark this answers is not in the document]
Comment (MJGaughran) [reply malcolm/modules/ADPandABlocks/parts/pandaseqtriggerpart.py]: Done

File: malcolm/modules/ADPandABlocks/doublebuffer.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+            0,
+            0,
+        )
+
+    @staticmethod
+    def _calculate_duration(rows: List[Tuple]) -> float:
+        """Return the duration of the given Sequencer rows (in list format)."""
+        duration = 0.0
+        for row in rows:
+            duration += row[0] * (row[3] + row[10])
```
[the remark this answers is not in the document]
Comment (MJGaughran) [reply malcolm/modules/ADPandABlocks/doublebuffer.py]: Now using namedtuples

**Review by thomascobb — APPROVED** [review]
I spent an hour reading it and what I understood looks fine. If there's anything in particular you'd like me to look at then let me know

**Pull request merged**